Libraries/BatchedBridge/BatchingImplementation/BatchedBridgeFactory.js

```diff
@@ -36,21 +44,40 @@ var BatchedBridgeFactory = {
    */
   _createBridgedModule: function(messageQueue, moduleConfig, moduleName) {
     var remoteModule = mapObject(moduleConfig.methods, function(methodConfig, memberName) {
-      return methodConfig.type === MethodTypes.local ? null : function() {
-        var lastArg = arguments.length > 0 ? arguments[arguments.length - 1] : null;
-        var secondLastArg = arguments.length > 1 ? arguments[arguments.length - 2] : null;
-        var hasSuccCB = typeof lastArg === 'function';
-        var hasErrorCB = typeof secondLastArg === 'function';
-        hasErrorCB && invariant(
-          hasSuccCB,
-          'Cannot have a non-function arg after a function arg.'
-        );
-        var numCBs = (hasSuccCB ? 1 : 0) + (hasErrorCB ? 1 : 0);
-        var args = slice.call(arguments, 0, arguments.length - numCBs);
-        var onSucc = hasSuccCB ? lastArg : null;
-        var onFail = hasErrorCB ? secondLastArg : null;
-        return messageQueue.call(moduleName, memberName, args, onFail, onSucc);
-      };
+      switch (methodConfig.type) {
+        case MethodTypes.remote:
+          return function() {
+            var lastArg = arguments.length > 0 ? arguments[arguments.length - 1] : null;
+            var secondLastArg = arguments.length > 1 ? arguments[arguments.length - 2] : null;
+            var hasErrorCB = typeof lastArg === 'function';
+            var hasSuccCB = typeof secondLastArg === 'function';
+            hasSuccCB && invariant(
+              hasErrorCB,
+              'Cannot have a non-function arg after a function arg.'
+            );
+            var numCBs = (hasSuccCB ? 1 : 0) + (hasErrorCB ? 1 : 0);
+            var args = slice.call(arguments, 0, arguments.length - numCBs);
+            var onSucc = hasSuccCB ? secondLastArg : null;
+            var onFail = hasErrorCB ? lastArg : null;
+            messageQueue.call(moduleName, memberName, args, onSucc, onFail);
+          };
+
+        case MethodTypes.remoteAsync:
+          return function(...args) {
+            return new Promise((resolve, reject) => {
```

Comment:
Here's a sample trace with bluebird's long stack traces enabled in JSC. Code looks like:

```js
  componentDidMount() {
    new Promise((resolve) => resolve(1)).then(() => {
      require('NativeModules').Testing.testAsync().catch(
        (err) => console.log(err.stack)
      );
    });
  }
```

```
Error: Unknown error from a native module
    _createErrorFromErrorData@UIExplorer:7330:24
    _invokeCallback@UIExplorer:7659:15
    forEach@[native code]
    perform@UIExplorer:6188:24
    batchedUpdates@UIExplorer:18840:26
    batchedUpdates@UIExplorer:4720:34
    applyWithGuard@UIExplorer:882:25
    guardReturn@UIExplorer:7487:30
From previous event:
    callTimer@UIExplorer:7981:17
    callImmediates@UIExplorer:8030:34
    _flushedQueueUnguarded@UIExplorer:7855:37
    applyWithGuard@UIExplorer:882:25
    guardReturn@UIExplorer:7490:37
From previous event:
    UIExplorerApp_componentDidMount@UIExplorer:1157:12
    notifyAll@UIExplorer:4957:26
    close@UIExplorer:19254:35
    closeAll@UIExplorer:6261:29
    perform@UIExplorer:6202:24
    batchedMountComponentIntoNode@UIExplorer:20834:22
    batchedUpdates@UIExplorer:18838:15
    batchedUpdates@UIExplorer:4720:34
    renderComponent@UIExplorer:20903:32
    ReactMount__renderNewRootComponent@UIExplorer:5056:26
    render@UIExplorer:1350:39
    renderApplication@UIExplorer:40446:15
    run@UIExplorer:40387:34
    runApplication@UIExplorer:40409:26
    jsCall@UIExplorer:7438:34
    _callFunction@UIExplorer:7693:21
    forEach@[native code]
    perform@UIExplorer:6188:24
    batchedUpdates@UIExplorer:18840:26
    batchedUpdates@UIExplorer:4720:34
    applyWithGuard@UIExplorer:882:25
    guardReturn@UIExplorer:7487:30
    processBatch@UIExplorer:7712:23
```

It's not clean but it's really good that it shows the error originated from componentDidMount. Compare it to the error with the current Promise library and no long stack traces:

```
Error: Unknown error from a native module
    _createErrorFromErrorData@UIExplorer:7328:24
    UIExplorer:7283:54
    _invokeCallback@UIExplorer:7657:15
    UIExplorer:7720:39
    forEach@[native code]
    UIExplorer:7712:22
    perform@UIExplorer:6186:24
    batchedUpdates@UIExplorer:13850:26
    batchedUpdates@UIExplorer:4718:34
    UIExplorer:7711:34
    applyWithGuard@UIExplorer:882:25
    guardReturn@UIExplorer:7485:30
    processBatch@UIExplorer:7710:23
    [native code]
    global code
```

So I think this will help contribute to much better debugging.